### ISSUE TITLE
Can't move pages

### ISSUE THREAD

**Issue by philippeowagner** [divio/django-cms#4436]
Can't move pages in admin. Throws an internal server error (500). 

    Traceback:
    File "/var/www/blank/www/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
      132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/django/utils/decorators.py" in _wrapped_view
      110.                     response = view_func(request, *args, **kwargs)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/django/views/decorators/cache.py" in _wrapped_view_func
      57.         response = view_func(request, *args, **kwargs)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/django/contrib/admin/sites.py" in inner
      233.             return view(request, *args, **kwargs)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/django/utils/decorators.py" in _wrapper
      34.             return bound_func(*args, **kwargs)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/django/views/decorators/http.py" in inner
      45.             return func(request, *args, **kwargs)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/django/utils/decorators.py" in bound_func
      30.                 return func.__get__(self, type(self))(*args2, **kwargs2)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/django/utils/decorators.py" in inner
      145.                     return func(*args, **kwargs)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/cms/admin/pageadmin.py" in move_page
      857.         page.move_page(target, position)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/cms/models/pagemodel.py" in move_page
      217.         moved_page = self.move(target, pos=position)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/cms/models/pagemodel.py" in move
      1237.         super(Page, self).move(target, pos)
    File "/var/www/blank/www/local/lib/python2.7/site-packages/treebeard/mp_tree.py" in move
      1037.         return MP_MoveHandler(self, target, pos).process()
    File "/var/www/blank/www/local/lib/python2.7/site-packages/treebeard/mp_tree.py" in process
      444.         if self.target.is_descendant_of(self.node):

    Exception Type: AttributeError at /admin/cms/page/250/move-page/
    Exception Value: 'NoneType' object has no attribute 'is_descendant_of'

Django Version: 1.8.4
Affected versions: django-cms==3.1.1 and django-cms==3.1.4.dev

**Comment by evildmp** [comment]
We need to be able to replicate this, preferably in a test case.

It sounds like MP tree corruption (really not what we want to hear, but still). 

Are you able to recreate the situation that causes this in a fresh database?

**Comment by philippeowagner** [comment]
It's still possible to move the pages below a certain level of depth. 
No, didn't tried with a fresh database.

**Comment by evildmp** [comment]
Does it seem to implicate a particular page as the problem?

**Comment by philippeowagner** [comment]
@evildmp I tried several things... Yes, it seems that the target page is the problem. What could help next?

**Comment by evildmp** [comment]
@philippeowagner I think we're going to need to see this in action, and probably to have a copy of the data in the page tree so we can replicate it - without being able to replicate it, we're in the dark. Is there some way we could arrange that?

**Comment by npardington** [comment]
I've also started getting this error on django-cms==3.1.3. 

Seems that treebeard->mp_tree.py line 442

    newdepth, siblings, newpos = self.update_move_to_child_vars()

Goes on to set self.target to None on line 529

    self.target = self.target.get_last_child()

That throws an exception on treebeard->models.py 277

    try:
        return self.get_children().reverse()[0]
    except IndexError:
        return None

I ended up deleting the page and recreating it to solve the problem. The structure it deleted seemed a bit odd 

![screen shot 2015-09-16 at 11 41 07 am](https://cloud.githubusercontent.com/assets/2708258/9910087/45ecf6c0-5c69-11e5-8be7-301983750af4.png)

**Comment by philippeowagner** [comment]
@evildmp we could try to reduce our client's DB to the CMS relevant thing and provide it in an example project. Would this help?

**Comment by philippeowagner** [comment]
I duplicated the target page and then could add all the children as expected. Had the same odd delete preview as @npardington reported. 

![screen shot 2015-09-16 at 20 09 40](https://cloud.githubusercontent.com/assets/113043/9913863/3de327d8-5caf-11e5-97f8-112f456252bc.png)

(reference to itself?)

**Comment by evildmp** [comment]
@npardington oh no, stop telling us this bad tree-related news! Do still have a copy of the bad data?

@philippeowagner yes please, the data in a suitably anonymised/depersonalised context would be great

**Comment by npardington** [comment]
I don't, but I did have to delete another page that had the same issue. If it comes up again I'll dump the data. 

Probably un related, but I have been seeing this error come up as well

    'NoneType' object has no attribute '_inc_path'

![screen shot 2015-09-16 at 4 26 18 pm](https://cloud.githubusercontent.com/assets/2708258/9917500/f29ac98a-5c8f-11e5-9afd-2c6300165d0e.png)

Haven't had a change to dive into that one, but the last time I looked it appeared to be a problem adding a link plugin inside the WYSIWYG.

**Comment by philippeowagner** [comment]
@npardington what happens when you delete the page? In my example it seem as it would remove the cloned page as well (and my example seems similar to yours....).

**Comment by npardington** [comment]
@philippeowagner yes when I deleted each page it had listed the cloned page as well.

**Comment by philippeowagner** [comment]
What to do to prevent the cloned page to be deleted as well? 

@evildmp we will clean a dump ...

**Comment by philippeowagner** [comment]
@evildmp FYA https://slack-files.com/T02BCAM2Z-F0AS1P7B3-e88fb1cea2

**Comment by npardington** [comment]
I'm using Postgres 9.3

**Comment by mkoistinen** [comment]
@philippeowagner, @npardington I had a slightly similar situation recently. I was able to fix it by running `python manage.py cms fix_tree`. HOWEVER, before you do this on a live project, please apply https://github.com/divio/django-cms/pull/4457 as a patch to CMS.

Could you do this with a ***copy*** of your DB and let us know if it helps?

**Comment by mkoistinen** [comment]
Also, when this happens, the CMS is probably providing a nice stack trace, but it is "hidden". To get to it (would be massively helpful), follow these steps:

1. In Chrome, open the dev-tools and click on the network tab;
2. Attempt to move the page, if this is still an issue, then the page won't move, but at the bottom of the network tab in dev tools, you should now see a red AJAX request, click this.
3. Click on the preview tab to see the stack trace.

Can someone provide this?

**Comment by philippeowagner** [comment]
@mkoistinen my stack trace was hidden, see first comment/issue.

**Comment by paperreduction** [comment]
@npardington thanks for the details, that helped me find this issue. I was having the same problem. @mkoistinen the fix-tree command resolved the data problem. The actual command I ran was `python manage.py cms fix-tree`. I'd love to know what cased the problem in the first place so we could get a real fix, but right now I honestly don't know what's happening under the hood with treebeard.

**Comment by paperreduction** [comment]
By the way, I have a copy of the database BEFORE applying the fix. So I can dump the records out for inspection. Would a SQL dump from postgres work?

**Comment by evildmp** [comment]
@paperreduction I should think so. But please make sure that it doesn't contain anything you don't want shared!

**Comment by evildmp** [comment]
I think enough people have reported this that we can accept it as a bug by the way.

**Comment by martcot** [comment]
python manage.py cms fix-tree works for me also.

**Comment by FinalAngel** [comment]
thank you for update @martcot 

@evildmp this should be resolved as ``python manage.py cms fix-tree`` is documented?

**Comment by philippeowagner** [comment]
@FinalAngel seriously? What should I tell the customer?

**Comment by FinalAngel** [comment]
I reformulated my answer on the top @philippeowagner. The goal is not to immediate resolve it via ``python manage.py cms fix-tree`` (which is a short-term fix) but rather find a proper solution.

https://github.com/divio/django-cms/pull/4457 will fix this issue in the 3.1.4 release. @yakky any comment on the release schedule there?

**Comment by philippeowagner** [comment]
@FinalAngel :+1:

**Comment by yakky** [comment]
@philippeowagner @FinalAngel @paperreduction @martcot  I'm focusin on fixing the treebeard issues. (linked)

**Comment by FinalAngel** [comment]
awesome @yakky !

**Comment by philippeowagner** [comment]
That's great @yakky ! Thank you !!

**Comment by npardington** [comment]
@yakky thanks for keeping up with this issue.

**Comment by martcot** [comment]
@yakky Thank you!

**Comment by yakky** [comment]
@philippeowagner @npardington @martcot would you try https://github.com/yakky/django-cms/tree/feature/treebeard?

**Comment by yakky** [comment]
@philippeowagner @npardington @martcot tentative fix at #4717

**Comment by philippeowagner** [comment]
@yakky will check ASAP with the data given earlier

**Comment by yakky** [comment]
@philippeowagner great, thanks

**Comment by philippeowagner** [comment]
@evildmp @yakky got the same error as reported 

````
Traceback:
File "/var/www/blank/dev/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/django/utils/decorators.py" in _wrapped_view
  110.                     response = view_func(request, *args, **kwargs)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/django/views/decorators/cache.py" in _wrapped_view_func
  57.         response = view_func(request, *args, **kwargs)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/django/contrib/admin/sites.py" in inner
  233.             return view(request, *args, **kwargs)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/django/utils/decorators.py" in _wrapper
  34.             return bound_func(*args, **kwargs)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/django/views/decorators/http.py" in inner
  45.             return func(request, *args, **kwargs)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/django/utils/decorators.py" in bound_func
  30.                 return func.__get__(self, type(self))(*args2, **kwargs2)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/django/utils/decorators.py" in inner
  145.                     return func(*args, **kwargs)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/cms/admin/pageadmin.py" in move_page
  861.         page.move_page(target, position)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/cms/models/pagemodel.py" in move_page
  217.         moved_page = self.move(target, pos=position)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/cms/models/pagemodel.py" in move
  1243.         super(Page, self).move(target, pos)
File "/var/www/blank/dev/local/lib/python2.7/site-packages/treebeard/mp_tree.py" in move
  1037.         return MP_MoveHandler(self, target, pos).process()
File "/var/www/blank/dev/local/lib/python2.7/site-packages/treebeard/mp_tree.py" in process
  444.         if self.target.is_descendant_of(self.node):

Exception Type: AttributeError at /admin/cms/page/250/move-page/
Exception Value: 'NoneType' object has no attribute 'is_descendant_of'
Request information:
GET: No GET data

POST:
position = u'last-child'
target = u'243'

FILES: No FILES data

COOKIES:
djangocms_nodes_open = 'page_203%2Cpage_49%2Cpage_112%2Cpage_33%2Cpage_243'
```

**Comment by yakky** [comment]
@philippeowagner did you run fix-tree before that? The PR fixes a tree corruption during a page delete thus any data must be fixed before testing the pr

**Comment by philippeowagner** [comment]
No @yakky first I did not running the fix-tree command. There are still broken element/pages in the dump that I restored. Also trying to delete shows the same effects as reported without running the command. 

After running `cmd fix-tree` everything works as expected for the page moving action, but the delete showed a even more wired result. 

![screen shot 2015-11-16 at 20 49 19](https://cloud.githubusercontent.com/assets/113043/11192835/8c5f36b2-8ca3-11e5-976c-fa9680663579.png)
